### ISSUE TITLE
Speculative contacts catch on seemingly inactive edges

### ISSUE THREAD

**Issue by mihe** [jrouwe/JoltPhysics#630]
Reported in godot-jolt/godot-jolt#508.

When running this:

```cpp
TriangleList triangles = {
    {Float3(-2.0f, 2.0f, 2.0f), Float3(2.0f, 2.0f, 2.0f), Float3(-2.0f, 2.0f, -2.0f)},
    {Float3(-2.0f, 2.0f, -2.0f), Float3(2.0f, 2.0f, 2.0f), Float3(2.0f, 2.0f, -2.0f)},

    {Float3(-2.0f, 2.0f, -2.0f), Float3(2.0f, 2.0f, -2.0f), Float3(-2.0f, 0.0f, -4.0f)},
    {Float3(-2.0f, 0.0f, -4.0f), Float3(2.0f, 2.0f, -2.0f), Float3(2.0f, 0.0f, -4.0f)},
};

BodyCreationSettings box(new BoxShape(Vec3(1, 1, 1)), RVec3(0, 3, 0), Quat::sIdentity(), EMotionType::Dynamic, Layers::MOVING);
box.mLinearVelocity = Vec3(0, 0, -0.5f);
box.mAngularDamping = 0.1f;
box.mLinearDamping = 0.1f;
box.mFriction = 0.0f;

mBodyInterface->CreateAndAddBody(box, EActivation::Activate);
mBodyInterface->CreateAndAddBody(BodyCreationSettings(new MeshShapeSettings(triangles), RVec3::sZero(), Quat::sIdentity(), EMotionType::Static, Layers::NON_MOVING), EActivation::DontActivate);

PhysicsSettings physics_settings = mPhysicsSystem->GetPhysicsSettings();
// physics_settings.mSpeculativeContactDistance = 0.0f;
mPhysicsSystem->SetPhysicsSettings(physics_settings);
```

You end up with this:

https://github.com/jrouwe/JoltPhysics/assets/4884246/ff066f6b-6c65-42cf-991c-3f868339f83d

But when setting the speculative contact distance to 0, you instead end up with this:

https://github.com/jrouwe/JoltPhysics/assets/4884246/7c4f2b19-74b0-4bf0-ac1a-c7789dc31fcc

I'm not sure if this is actually related to speculative contacts, or active/inactive edges for that matter, but there you have it.

**Comment by jrouwe** [comment]
I'm AFK this week, so I will look at this next week.

**Comment by mihe** [comment]
I just realized this exact scenario was already being tested in the `ActiveEdgesTest` and I had somehow overlooked the boxes sliding down the convex slopes.

And indeed, if you change this velocity in `ActiveEdgesTest` to `Vec3(0, 0, 0.5f)` the same issue appears:

https://github.com/jrouwe/JoltPhysics/blob/b1cd84951022a565a23f33ff77ffee9e71487bbc/Samples/Tests/General/ActiveEdgesTest.cpp#L83-L85

https://github.com/jrouwe/JoltPhysics/assets/4884246/f465dd9f-279f-4ada-b3ba-5e2c1e6ddb2e

Although, I'm realizing now, after enabling the "Draw Mesh Shape Triangle Outlines" drawing option, that these are in fact not inactive edges, because the two faces are angled apart by more than the [5 degree cut-off](https://github.com/jrouwe/JoltPhysics/blob/b1cd84951022a565a23f33ff77ffee9e71487bbc/Jolt/Physics/Collision/ActiveEdges.h#L30). The video compression muddles up the colors a bit, but the edges for the triangles where the boxes come to a stop are all colored green, indicating an active edge.

Either way, the test itself seems to imply that this scenario should still work, so presumably something is amiss.

Just thought I'd mention that. Enjoy your AFK time!

**Comment by jrouwe** [comment]
So this is a known issue, you don't have to modify the ActiveEdgesTest at all to see it, just run it and some of the blocks on the slopes going down will hit an active edge. The current active edge implementation is quite simple:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/50021b85-44ac-411b-b301-7c89226d532b)

So the edge in your example is indeed active and the block can collide with it. Even when setting the speculative contact distance to zero, this will still happen, it will only make it less likely.

The reason that there is a cut off angle can be seen in this picture:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/8496ad00-3060-43c6-8ceb-5de7fc1b809b)

The red sphere hits a very sharp edge, the deepest penetration (length and direction) is indicated by the green arrow. What inactive edges do is that they take the deepest penetration distance but use the face normal as direction (blue arrow). This works well if the edge is not sharp, but it doesn't work well in the example above as the sphere will not be fully pushed out (the distance along which it needs to be pushed is way bigger in this direction).

185 degrees is maybe a bit too conservative, I would expect most edges up to 270 degrees to work fine with the current strategy. Another option would be to have some run-time edge detection code. This needs some further thought and time.

**Comment by jrouwe** [comment]
Hello,

I added an option to MeshShapeSettings to override the threshold angle. I don't dare to change the default angle at this point, but I think you can easily set the angle to 50 degrees without introducing artifacts. This should fix this particular test.

**Comment by mihe** [comment]
Great, I'll give that a try. Thank you!

**Comment by Gamepro5** [comment]
> The red sphere hits a very sharp edge, the deepest penetration (length and direction) is indicated by the green arrow. What inactive edges do is that they take the deepest penetration distance but use the face normal as direction (blue arrow). This works well if the edge is not sharp, but it doesn't work well in the example above as the sphere will not be fully pushed out (the distance along which it needs to be pushed is way bigger in this direction).

Pardon me for not being smart, but what is the downside to always pushing it out in the direction of the green arrow?

Also, correct me if I'm wrong, but what I'm gathering is that this isn't an issue that can really be solved and the current workaround is playing around with the threshold angle? There is no "perfect" angle that solves all edge cases (pun not intended). If this is the case, are there any potential true fixes?

**Comment by jrouwe** [comment]
The downside of pushing it out in the direction of the green arrow is that in your case the green arrow (from the intersection between the box and the edge of the sloped part) is pointing towards the box and is stopping the box:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/f61b0e92-8983-41c0-94a3-b875a66cc61a)

(note that the box sinks into the ground a little bit but not as much as drawn here)

You're completely right that there is no perfect angle, but I think this fix will allow you to get rid of most if not all problem cases. In order to fix all cases you would need to detect that you can ignore this edge at run-time because the box is resting on the horizontal plane (has an interior hit on the face connected to the edge of the slope that we're hitting):

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/80a35bbc-5bd4-497c-8a73-d6b02434f10c)

(note here green arrow is velocity, not contact normal)

This involves looping over all contact manifolds and comparing them with other contact manifolds so comes at a performance cost, and since we never ran into this issue I didn't want to waste any CPU cycles on this.

**Comment by Gamepro5** [comment]
> This involves looping over all contact manifolds and comparing them with other contact manifolds so comes at a performance cost, and since we never ran into this issue I didn't want to waste any CPU cycles on this.

Would it be possible to have a setting to turn this on? For simulations, obviously every CPU cycle counts and if there are a hundreds of shapes being simulated, I can understand this optimization. However, when using it for a character controller, I would like to ensure 100% accuracy. (In Godot Jolt, a lot of issues I've encountered with move_and_collide() for character controllers have been directly related to this)

**Comment by mihe** [comment]
> In Godot Jolt, a lot of issues I've encountered with move_and_collide() for character controllers have been directly related to this.

I had hoped to build a better understanding of all of this before starting this discussion, but there's a lot of stuff to unpack here, so here's some context...

`move_and_collide` is for all intents and purposes [this](https://github.com/godot-jolt/godot-jolt/blob/f702c0fc30d4971836296530415082b7f4a24080/src/spaces/jolt_physics_direct_space_state_3d.cpp#L499-L570) method in Godot Jolt, which in turn is a port of [this](https://github.com/godotengine/godot/blob/b51ee8b029b0b9f719f01bbdd21a329e65d4d238/servers/physics_3d/godot_space_3d.cpp#L650-L1028) method in Godot Physics, and can roughly be considered to be the Godot equivalent of `CharacterVirtual::MoveShape` I guess. This method is largely intended to be used by character controllers, and is the only part of Godot's `CharacterBody3D` that a physics implementation is responsible for implementing.

My implementation of this is currently using `EActiveEdgeMode::CollideOnlyWithActive` for all the various `CollideShape` calls that are involved in it, meaning the iterative displacement (called "recovery" in Godot), the sweep/cast (binary search) and the final collision check. This seems to have a mostly-ish positive impact on Godot's `CharacterBody3D`, or to be more specific its `move_and_slide` method (found [here](https://github.com/godotengine/godot/blob/b51ee8b029b0b9f719f01bbdd21a329e65d4d238/scene/3d/physics_body_3d.cpp#L1164-L1509)), but is seemingly having a negative impact on @Gamepro5's character controller, because it (from what I understand) needs the (average?) contact normal to not be messed with by the active edge detection.

Switching to using `EActiveEdgeMode::CollideWithAll` instead seems to have mostly resolved that problem for @Gamepro5 (correct me if I'm wrong), and could be a candidate for a new project setting in Godot Jolt, but it (perhaps unsurprisingly) results in a couple of regressions where you'll sometimes end up getting some unexpected contact normals when walking across certain triangle edges, which I assume are the same "ghost collisions" that `EActiveEdgeMode::CollideOnlyWithActive` and this active edge threshold is meant to address in the first place.

There have also been some instances where regardless of what `mActiveEdgeMode` is set to, coupled with this increased active edge threshold of 50 degrees, you still end up with some unexpected normals, but I haven't had the time to dig into why exactly that is. From what I understand it's possible for Jolt to consider certain internal edges to still be active depending on how the mesh is triangulated, so maybe that's what's happening there.

Anyway, I guess the assumption here is that this proposed run-time edge detection would resolve 100% of all ghost collisions while also not affecting the normals in the same way as `EActiveEdgeMode::CollideOnlyWithActive`, which strikes me as a tall order, but maybe I'm wrong.

I'm also not entirely convinced yet that this couldn't be resolved on my side somehow. There are still some minor discrepancies between my implementation of `move_and_collide` and the one found in Godot Physics. For example, Godot Physics discards contacts whose depth is below a certain calculated threshold, whereas I don't, and from what I've seen a lot of the "problematic" contacts are typically quite shallow compared to the expected ones, so maybe that could help.

**Comment by Gamepro5** [comment]
> because it (from what I understand) needs the (average?) contact normal to not be messed with by the active edge detection.

First of all, I should mention that my specific character controller isn't really relevant here. It is an alternative to `move_and_slide()` that I've been working on for a few years as when I started, `move_and_slide()` was bugged.

The average normal you quizzically mentioned is to resolve situations where all collision normals are considered "too steep" because the player is wedged in a V shape. I therefore add up all the normals I'm colliding with and normalize it each time to prevent the player from getting stuck in the V shape. (This may be outdated, but I made [this blog](https://gamepro5.com/teamspec/devlog/Character_Controller_Physics/Character_Controller_Physics) explaining this. The graphic there still illustrates my point for the average normals)

> Switching to using EActiveEdgeMode::CollideWithAll instead seems to have mostly resolved that problem for @Gamepro5 (correct me if I'm wrong)

It improved some things and broke other things. MRPs for the regression is [here](https://github.com/godot-jolt/godot-jolt/issues/530#issuecomment-1681410669), and the thing that it didn't fix (but wasn't a regression) is [here](https://github.com/godot-jolt/godot-jolt/issues/530#issuecomment-1681479488). From my non professional limited testing, it seems to be "not worse", but also "not fixed".

> Anyway, I guess the assumption here is that this proposed run-time edge detection would resolve 100% of all ghost collisions while also not affecting the normals in the same way as EActiveEdgeMode::CollideOnlyWithActive, which strikes me as a tall order, but maybe I'm wrong.

I'm nowhere near as experienced as either of you in low level collision detection, but from [jorit's explanation](https://github.com/jrouwe/JoltPhysics/issues/630#issuecomment-1683939782), I interpreted that `"In order to fix all cases you would need to detect that you can ignore this edge at run-time because the box is resting on the horizontal plane (has an interior hit on the face connected to the edge of the slope that we're hitting)"` was describing how it could be done, but that he didn't want to waste precious processing time on it because most of the time it doesn't really matter.

**Comment by jrouwe** [comment]
> Would it be possible to have a setting to turn this on? For simulations, obviously every CPU cycle counts and if there are a hundreds of shapes being simulated, I can understand this optimization.

It is something that is on my list, but is unfortunately not trivial and I don't know for sure that it will turn out to be better.

We new went from a box sliding to a ramp to a character, so I'm having a bit of trouble to understand what's going on now. I have the idea that what I tried to solve initially is not actually the problem. I'm having a hard time understanding what @Gamepro5's character does and how this maps to @mihe's implementation of move_and_collide and how that function exactly works. @mihe are you using are you using CollideShapeSettings::mActiveEdgeMovementDirection and CollideShapeSettings::mBackFaceMode in that function?

**Comment by Gamepro5** [comment]
My character controller is a box that slides on ramps. Its specifics don't really matter, all that is relevant is that it uses a `move_and_collide()` which apparently is `JoltPhysicsDirectSpaceState3D::test_body_motion`. I then use the normals it reports to properly move the character around. The issue comes from the fact that even a small occasional error in normals is enough to suddenly make the player stop, and I simplified the issue as a single `JoltPhysicsDirectSpaceState3D::test_body_motion` call.

**Comment by mihe** [comment]
> We new went from a box sliding to a ramp to a character, so I'm having a bit of trouble to understand what's going on now. I have the idea that what I tried to solve initially is not actually the problem.

The configurable threshold you provided seems to have [worked great](https://github.com/godot-jolt/godot-jolt/issues/508#issuecomment-1677079674) as far as dynamic bodies are concerned, and it also happened to resolve (or at least alleviate) [another issue](https://github.com/godot-jolt/godot-jolt/issues/524) related to `CharacterBody3D` and its `move_and_slide` method (which is in part powered by `move_and_collide`).

This `move_and_collide` thing might warrant a new discussion/issue, rather than re-opening this one, but with all the different parts involved in this I honestly don't have a good enough sense of the actual problem yet to be able to point to any specific thing as the cause, so I would also be fine with continuing this discussion once I do. While toggling `mActiveEdgeMode` seems to have shuffled things around a bit, it clearly doesn't resolve all discrepancies with Godot Physics one way or the other.

> [...] @mihe's implementation of move_and_collide and how that function exactly works.

It's a lot, unfortunately, and I'm sure I will end up omitting some important details, but in very broad terms:

1. Resolve any current overlaps using `CollideShape`, with a small `mMaxSeparationDistance` (usually 0.001) added to it. This also gets added to any resulting penetration depth/axis, since that seems to get subtracted within Jolt. This is the `_body_motion_recover` part of `test_body_motion`.
2. Perform a sweep/cast (binary search) from this resolved position along the motion vector, without using any `mMaxSeparationDistance`. This is the `_body_motion_cast` part of `test_body_motion`.
3. Perform a `CollideShape` at the "unsafe" fraction returned by that sweep, with a `mMaxSeparationDistance`, which again gets added to any resulting penetration depth. These are the actual contacts returned to the user and is the `_body_motion_collide` part of `test_body_motion`.
4. (Optional) Actually move the body by the displacement from step 1 and the "safe" fraction from step 2.

> @mihe are you using are you using CollideShapeSettings::mActiveEdgeMovementDirection and CollideShapeSettings::mBackFaceMode in that function?

I am using `mActiveEdgeMovementDirection`, yes, which is set to the normalized motion vector. I've tried omitting this one a couple of times, but never really noticed any difference, so its practical effects are a bit of a mystery to me still.

`mBackFaceMode` is left at the default value of `IgnoreBackFaces`. As far as I can tell Godot Physics ignores back-faces too, so I've never really messed with it. Toggling it seems to shuffle some things around, where certain triangle edge collisions (in the context of `move_and_slide`) goes away, but others are still present. It's tough to say whether there's any improvement.

**Comment by jrouwe** [comment]
Ok, mActiveEdgeMovementDirection and mBackFaceMode look good. The only use for mActiveEdgeMovementDirection is that when we hit an edge and the calculated normal affects movement less than the triangle normal, we will always use the calculated normal (regardless if the edge is active or not). This was added to help objects slide over edges.

I need to chew on this, especially in combination with #671. 

We've always used a capsule shape for the character because a capsule is just so much easier to slide around the world than a box shape (a box shape is much more sensitive to the contact normal, and a capsule is much easier for collision detection than a box because it's a line segment with a large convex radius). Jolt's VirtualCharacter also tries to avoid all of these issues by introducing an artificial convex radius around the character (using mMaxSeparationDistance) but actually keeping the character away from the surface by this distance so that float precision issues are minimized.

**Comment by mihe** [comment]
> Jolt's VirtualCharacter also tries to avoid all of these issues by introducing an artificial convex radius around the character (using mMaxSeparationDistance) but actually keeping the character away from the surface by this distance so that float precision issues are minimized.

That's how `move_and_collide` (and `move_and_slide` as a result) is meant to work as well, for similar reasons. Godot calls this the ["safe margin"](https://docs.godotengine.org/en/stable/classes/class_characterbody3d.html#class-characterbody3d-property-safe-margin), and is the `mMaxSeparationDistance` that I mentioned above that defaults to 0.001. This effectively makes the body/shape hover over the ground by that amount. The way it's actually incorporated might be different though.

**Comment by jrouwe** [comment]
I've been thinking of this one: I can update the simulation to properly detect active edges, but it won't actually help for the ```move_and_collide``` function because that is a different code path. 

In order to make this function properly detect active we would have to use the ```CollideShape``` function with ```CollideShapeSettings::mCollectFacesMode = ECollectFacesMode::CollectFaces``` and collect _all_ hits.

For each hit, we would need to check if the hit is close to an edge (of CollideShapeResult::mShape2Face). If it is then check all other hits to see if the edge overlaps at the contact point with another edge. If so, check if the other hit is away from the edge, and if it is discard the first hit.

To illustrate with our example:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/ea53c0d1-b6f0-41d1-b912-e959308d48bd)

Triangle 1 and 2 both collide with the box, but triangle 1 has a hit on the edge while triangle 2 has a hit that is in the interior of the triangle so it is away from the edge between triangle 1 and 2 so we can discard the hit with triangle 1.

I can provide this as a utility method, but I'm wondering @Gamepro5 if after [this](https://github.com/jrouwe/JoltPhysics/commit/12adfda9621f3f568b7c68a665d718fffa1d701c) and [this](https://github.com/godot-jolt/godot-jolt/commit/342b8ea61341adf6e6159fd792aaa48fa89e97c3) fix, the issue is now mostly theoretical or if it still actually happening?

**Comment by mihe** [comment]
> if after [this](https://github.com/jrouwe/JoltPhysics/commit/12adfda9621f3f568b7c68a665d718fffa1d701c) and [this](https://github.com/godot-jolt/godot-jolt/commit/342b8ea61341adf6e6159fd792aaa48fa89e97c3) fix, the issue is now mostly theoretical or if it still actually happening?

Having taken a closer look at the various repro projects, some of which ended up not being strictly related to this issue, the actual problem here essentially boils down to only wanting the green normals here, rather than any of the red normals, with the yellow arrow indicating the direction of movement for the box:

![image](https://github.com/jrouwe/JoltPhysics/assets/4884246/dc8402bd-1ae2-472d-a530-10636bd9132a)

![image](https://github.com/jrouwe/JoltPhysics/assets/4884246/c0b50629-1b61-45e9-a7c9-2eab1e9d5833)

In the first image/scenario, you will currently end up with normal A because of the normal being overridden by `CollideOnlyWithActive`, which has now been made "worse" in Godot Jolt by me setting the threshold to 50 degrees. Switching over to `CollideWithAll` technically fixes this particular repro, but I would imagine that normal B (or something like it) is still a possibility with `CollideWithAll` depending on the placement of the box, so I suspect this doesn't actually resolve anything.

In the second image/scenario, the box is placed right alongside a triangle edge, leading to contacts with normals C and D. When using `CollideOnlyWithActive` it actually corrects normal C a tiny bit as it's being overridden with the face normal, but because of `mActiveEdgeMovementDirection` pointing straight down, and a leftward-facing direction being less obstructing than the upward-facing face normal, it picks the leftward-facing one instead.

Switching to `CollideWithAll` in that second scenario obviously does nothing to fix anything, since you'll just end up with largely the same normals. Omitting the `mActiveEdgeMovementDirection` does however fix this, but this would mean leaving `CollideOnlyWithActive` enabled in the first place, which clashes with scenario 1.

All-in-all, I struggle to see a permutation of settings here that will satisfy every edge-case. While there is some argument for a project setting in Godot Jolt that enables `CollideWithAll` instead, seeing as how that's technically a discrepancy with how Godot Physics behaves, you will undoubtedly end up with other awkward edge normals instead, much like you do with Godot Physics already, albeit less so.

And as mentioned already, `move_and_slide`, which uses `move_and_collide`, and is effectively what constitutes Godot's character controller, benefits greatly from `CollideOnlyWithActive`, so I can't help but feel like dealing with the normals it outputs from the character controller instead shouldn't be an insurmountable obstacle.

**Comment by jrouwe** [comment]
What about the algorithm I suggested? (as far as I can see it will fix both cases)

**Comment by mihe** [comment]
> What about the algorithm I suggested? (as far as I can see it will fix both cases)

It seems I got ahead of myself and glossed over it. It's certainly worth a shot, now that I'm reading it more carefully.

I'm assuming this would be more expensive than just a simple hit collector then, seeing as how it has to collect faces and all this extra stuff? Would it make sense then to gate this behavior behind some sort of project setting as well, or perhaps just leave the regular shape queries out of this and let this algorithm be how the kinematic shape queries (i.e. `move_and_collide`) behave?

Either way, the details of the implementation of this are a bit hazy to me still, so if you feel a utility function like this makes sense in Jolt, then that'd be much appreciated.

**Comment by jrouwe** [comment]
It is indeed a lot more expensive so I would use it on ```move_and_collide``` only.

I'll see if it is sensible as a utility and otherwise I'll paste the code here.

**Comment by Gamepro5** [comment]
I don't know if this is possible to do with GDNative, but maybe a parameter for `move_and_collide()` could be added to opt into this algorithm if it really is that much more expensive.

If the default `move_and_slide()` (which is a higher level implementation of a character controller using `move_and_collide()`) is forced to handle these edge cases in code, I feel like that maybe less efficient than doing it directly in the physics engine. `move_and_slide()` probably needs to do a bunch of other physics server requests to ensure the normals being given are reasonable.

**Comment by mihe** [comment]
I made a naive/inefficient attempt at the proposed algorithm [here](https://github.com/mihe/godot-jolt/blob/b085058c3d720ea10f8b123ed11ca64b341e548b/src/spaces/jolt_physics_direct_space_state_3d.cpp#L939-L990). Is that roughly what you had in mind, @jrouwe?

It does seem to resolve the various repros related to this, but I'm also seeing some regressions/jank with `move_and_slide` when transitioning from a slope/ramp to a flat surface, which maybe isn't too surprising considering how this works right now, but I figured I should check to see whether I'm on the right track.

**Comment by jrouwe** [comment]
Sorry somehow missed the notification yesterday. I was working on the same thing yesterday afternoon and realized that the algorithm needs to be a bit more complicated than I described. I'm hoping to have something that works today.

**Comment by jrouwe** [comment]
I've created a prototype here:

https://github.com/jrouwe/JoltPhysics/tree/feature/runtime_welding

It needs further testing to see if it works in all cases, but you could try it out (in combination with setting the convex radius > 0).

Note that the class requires some specific settings on the CollideShapeSettings as documented here:

https://github.com/jrouwe/JoltPhysics/blob/bbbb95d0c780622188e657f4534ce497c36330e6/Jolt/Physics/Collision/WeldingCollideShapeCollector.h#L16-L17

**Comment by mihe** [comment]
Some observations from some quick preliminary tests:

1. You seemingly forgot to include `QuickSort.h`
2. Judging by the output from this, and from the code itself, I assume this will still output normals similar to normal A described in scenario 1 [here](https://github.com/jrouwe/JoltPhysics/issues/630#issuecomment-1694669208), albeit averaged between the two surface normals?
3. One of the repros for this consists of a box colliding with two separate convex hull shapes (see illustration below), which this doesn't seem to do anything about.
4. While the averaged surface normals produced by this seems to lessen the jank that I was seeing from my own naive attempt, it doesn't completely eliminate it. When using this with a regular capsule character and `move_and_slide` you get a noticeable snap/jank as you walk up a slope and transition to horizontal ground, presumably because there's no smooth progression of normals at the edge anymore. While it's certainly possible to just gate this behavior behind a project setting, I had hoped that this could be a more general-purpose solution, but maybe that's not realistic.

I can throw together a `Samples` test that repros the various problems described here, but I figured I'd hear your thoughts first.

![image](https://github.com/jrouwe/JoltPhysics/assets/4884246/9c3f04d7-dafa-4fe1-b926-af7f5693d447)

**Comment by jrouwe** [comment]
> You seemingly forgot to include QuickSort.h

Yes I did, I was in a hurry because I had to leave. I see I made a typo in the algorithm too.

> Judging by the output from this, and from the code itself, I assume this will still output normals similar to normal A described in scenario 1 https://github.com/jrouwe/JoltPhysics/issues/630#issuecomment-1694669208, albeit averaged between the two surface normals?

For scenario 1 there are multiple cases:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/56a3633f-9782-449a-88d6-5123a5bc517a)

Then this one:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/1776f6d7-ebf7-4b70-acc0-cf2a71771611)

And finally this one:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/03394409-ef26-42c6-9325-047e434886b9)

Unfortunately it also comes with this one:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/4b997b2c-784c-4e04-95f6-43ad12ae2b60)

> One of the repros for this consists of a box colliding with two separate convex hull shapes belonging to the same body (see illustration below), which this doesn't seem to do anything about.

I'll investigate this one. 

> While it's certainly possible to just gate this behavior behind a project setting, I had hoped that this could be a more general-purpose solution, but maybe that's not realistic.

A character mover is one of the most difficult things to do. We're trying to infer what the world looks like based on a few contact points for which we got imperfect contact positions (because the contact isn't a point at all) and contact normals. In this case you don't even know that you're testing for a character or what 'up' is, so that severely limits the options for cheating (which CharacterVirtual definitely does).

**Comment by mihe** [comment]
> For scenario 1 there are multiple cases

Right, so if I understand you correctly then this can be roughly summed up as "override the normals of edge contacts with the face normal of any neighboring contact when possible, and if not, average the two face normals"?

Judging by that description, as well as the behavior of another relevant repro here, I'm guessing then that this becomes somewhat susceptible to how the two planes on this slope are tessellated?

This is that repro:

https://github.com/jrouwe/JoltPhysics/assets/4884246/467de40d-8408-4ca3-8c33-fa045b31e385

... where you end up with almost every variety of the cases you presented, and some contacts (the lower ones) being "incorrectly" welded with neighboring faces on the same plane. I guess the ideal outcome here was for this to only ever give you normals on the vertical plane, but I assume that would be tough with this technique, unless you do some kind of iterative welding.

If the normals generated by `CollideOnlyWithActive` (which look very similar in this case) was a problem then I imagine this would be too.

> A character mover is one of the most difficult things to do. We're trying to infer what the world looks like based on a few contact points for which we got imperfect contact positions (because the contact isn't a point at all) and contact normals. In this case you don't even know that you're testing for a character or what 'up' is, so that severely limits the options for cheating (which CharacterVirtual definitely does).

Fair enough. I'm mostly worried that the effort put into of all of this will only ever be applicable to a very specific (or even singular) use-case, but I guess that's just how things go sometimes.

Anyway, one thing I forgot to bring up is that I currently use [my own collectors](https://github.com/godot-jolt/godot-jolt/blob/master/src/spaces/jolt_query_collectors.hpp) rather than the ones provided by Jolt, mostly in order to avoid needing to allocate memory for the `std::vector` of hits. The cost of doing so might very well pale in comparison to the rest of the work being done in this particular context, but I figured if you saw some way of implementing this as utility functions instead, or if you don't mind me copying the collector wholesale into Godot Jolt once it's done, then that would jive better with how things are done currently.

**Comment by jrouwe** [comment]
> This is what that repro looks like when using WeldingCollideShapeCollector

I think this is a reasonable set of normals, based on the limited information that is available, I wouldn't expect there to be a much better set. The main thing the algorithm does is to prevent normals like the blue ones (hard to draw them, see them as being in the plane of the triangle they originate from):

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/5d417951-dab7-4c6b-ad63-fa615faad022)

Those are the kind of normals that really affect the simulation.

> Anyway, one thing I forgot to bring up is that I currently use [my own collectors](https://github.com/godot-jolt/godot-jolt/blob/master/src/spaces/jolt_query_collectors.hpp) rather than the ones provided by Jolt

Yes I was aware of that. I wanted to get things working first, since it is in a header, it is easy to turn this into a template.

**Comment by jrouwe** [comment]
> One of the repros for this consists of a box colliding with two separate convex hull shapes belonging to the same body (see illustration below), which this doesn't seem to do anything about.

I created a repro of this:

```
void StaticCompoundShapeTest::Initialize()
{
	// Floor
	CreateFloor();

	// Create box
	Array<Vec3> box;
	box.push_back(Vec3(5, 0.5f, 5));
	box.push_back(Vec3(-5, 0.5f, 5));
	box.push_back(Vec3(5, -0.5f, 5));
	box.push_back(Vec3(-5, -0.5f, 5));
	box.push_back(Vec3(5, 0.5f, -5));
	box.push_back(Vec3(-5, 0.5f, -5));
	box.push_back(Vec3(5, -0.5f, -5));
	box.push_back(Vec3(-5, -0.5f, -5));

	// Simple compound
	Ref<StaticCompoundShapeSettings> compound_shape = new StaticCompoundShapeSettings;
	compound_shape->AddShape(Vec3(0, 5.5f, 0), Quat::sIdentity(), new ConvexHullShapeSettings(box, 0.01f));
	compound_shape->AddShape(Vec3(5.5f, 1.0f, 0), Quat::sRotation(Vec3::sAxisZ(), 0.5f * JPH_PI), new ConvexHullShapeSettings(box, 0.01f));

	Body &body = *mBodyInterface->CreateBody(BodyCreationSettings(compound_shape, RVec3::sZero(), Quat::sIdentity(), EMotionType::Static, Layers::NON_MOVING));
	mBodyInterface->AddBody(body.GetID(), EActivation::Activate);
}
```

And it indeed returns the wrong normal:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/244daabe-8185-4e14-827e-57af9cff7f5d)

But if you zoom in closer you can see why:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/e3b24f72-8616-4749-98c1-af568288e3fa)

The convex radius of the convex hull results in a contact point that is a little bit away from the edge (because the edge is rounded), so the algorithm doesn't see this as a connected edge and doesn't fix the normal. Increasing ```inOnEdgeToleranceSq``` or reducing the convex radius fixes it:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/308fe2c9-fbeb-4c82-9390-592cfff12e02)

So it all boils down to tweaking epsilons (as usual) 😭. GJK/EPA knows very little about the shape it's colliding, so I don't see how this could be made much better unless a connectivity map is made beforehand, storing the neighbors of each face (= lots of memory and expensive to maintain).

**Comment by mihe** [comment]
> I think this is a reasonable set of normals, based on the limited information that is available, I wouldn't expect there to be a much better set.

In that case we find ourselves at a strange junction. These kinds of normals were largely what spawned the issue that in turn rebooted this issue in the first place. I feel weird arguing for this since I have no issues with them myself, and neither does `move_and_slide` from what I can tell, but if this is the best-case scenario then frankly using a regular collector along with `CollideOnlyWithActive` and just omitting `mActiveEdgeMovementDirection` would presumably have a very similar success rate, since the only repro that `WeldingCollideShapeCollector` ends up actually "fixing" (by some arbitrary measure) is one caused by `mActiveEdgeMovementDirection`.

**Comment by jrouwe** [comment]
I think the big difference between ```CollideOnlyWithActive``` and this new method that does the classification at run time is:

* Ability to weld across compound shapes / non-mesh shapes.
* Doesn't include an arbitrary welding angle (```mActiveEdgeCosThresholdAngle```) or movement direction (```mActiveEdgeMovementDirection```).
* Can bend the normal to be the average between 2 faces instead of only having 1 face normal available.

I have to admit I'm running out of ideas here.

**Comment by jrouwe** [comment]
I changed the class so you can provide your own container class.

**Comment by mihe** [comment]
Thank you, but just so there's no misunderstanding here, `WeldingCollideShapeCollector` in its current form is not something I can really make use of. Maybe it makes sense to bring into Jolt anyway, as an alternative to `CollideOnlyWithActive`, but don't feel the need to accommodate for my use of it in that case.

**Comment by jrouwe** [comment]
Ok, but does that mean you're going to stick with what you had for move_and_collide? And if so, what shall we do with this ticket?  

B.t.w. I'm fine if the code won't get used but then I'll probably discard it as I don't have a real use case for it either at the moment.

**Comment by mihe** [comment]
> Ok, but does that mean you're going to stick with what you had for move_and_collide?

Yes, there's not much else I can do really. `WeldingCollideShapeCollector` isn't applicable to `move_and_slide`, which I would argue is the primary use-case for `move_and_collide`, and it also doesn't seem to move the needle much as far as the various repros provided by @Gamepro5 are concerned.

> And if so, what shall we do with this ticket?

We seem to have taken this to some kind of logical conclusion, so I don't have any problems with closing the ones on my end as "won't/can't fix". Maybe there's still hope for #671 being resolved in some way, but the question is if this issue remaining unresolved is a dealbreaker for @Gamepro5 in terms of using Jolt, in which case spending more time on that one might be moot as well.

**Comment by jrouwe** [comment]
Let's see what @Gamepro5 thinks.

**Comment by Gamepro5** [comment]
First of all, I want to thank you both for all the effort put into attempting to fix this. I don't understand half the things you are saying, collision detection is way too low level for me.

This not being resolved is in no way a dealbreaker, as there isn't really much else I can use. While godot physics doesn't have these particular issues, it is unstable in countless other ways.

From what I understand this issue is still valid. I don't think it should be closed. Someone might come along someday that knows how to fix this.

As for me, I really want to try to figure out why mihe says `move_and_slide()` doesn't encounter these issues. It's very odd how `move_and_collide()` reports them but `move_and_slide()` doesn't, as Godot simply uses `move_and_collides()` to make `move_and_slide()` work. I tried looking at the source code for `move_and_slide()` and it isn't really obvious to me why this issue doesn't appear.

I do know that a non zero shape margin makes none of these issues appear. It also would appear as though shape margins are nowhere near as detrimental to normal calculation as they were during the brief period of time I tried using a capsule collision shape got the idea from here (https://thelowrooms.com/articledir/programming_stepclimbing.php) but the problem with that one was [this](https://github.com/godot-jolt/godot-jolt/issues/471 ), which also happens with godot physics. I figured a box shape was my only real option for a quake like character controller. Additionally, when using shape margins with the cylinder, it would behave kind of like a capsule near the edges and give rounded normals, so I would kind of round off on a slope. This doesn't seem to be the case with boxshape3d, so I'm wondering if I've just been blind to the simple solution of using the shape margins for boxshape3d.

I haven't done much testing with move_and_slide(), so either it shares the same issue but we haven't found it, or move_and_slide() does something to curate the normals received from the `move_and_collides()`.

If you are willing to keep trying on https://github.com/jrouwe/JoltPhysics/issues/671, I'm down to keep testing, but if you have bigger bugs to fry, then I don't think it's worth the time considering what I just said in the above paragraph.

Thanks again for your professionalism and dedication to solving this problem!

**Comment by jrouwe** [comment]
> First of all, I want to thank you both for all the effort put into attempting to fix this.

And thank you for all your time creating repro's 😸 (and of course @mihe for all of his work!)

> From what I understand this issue is still valid. I don't think it should be closed.

I'd prefer to close tickets that are not going to be actively worked on (it makes my todo list easier), it will remain accessible for everyone and we can always reopen it if someone comes along.

> During the brief period of time I tried using a capsule collision shape got the idea from here (https://thelowrooms.com/articledir/programming_stepclimbing.php) but the problem with that one was https://github.com/godot-jolt/godot-jolt/issues/471

I took a quick look. I'm guessing that the main problem is that godot doesn't expose enough information about the collision to do proper reasoning. If you look at Jolt's character implementation:

https://github.com/jrouwe/JoltPhysics/blob/5d0838e48527a1344e3d3aecea933c043879019f/Jolt/Physics/Character/CharacterVirtual.cpp#L397-L398

You'll see that it uses the surface normal (the thing you want) to determine if a slope is too steep, but it uses the contact normal (```mContactNormal```, which is the thing you get through godot-jolt right now) for other things.

It's a shame that ```PhysicsServer3DExtensionMotionCollision``` (the class that stores the godot collision information) doesn't have a place to store a ```SubShapeID``` (Jolt's way of tracking what got hit) because then a custom function could be exposed to extract the surface normal for a collision.

**Comment by jrouwe** [comment]
Backing up the ```WeldingCollisionShapeCollector``` in case it is ever needed again.

[WeldingCollideShapeCollector.zip](https://github.com/jrouwe/JoltPhysics/files/12516478/WeldingCollideShapeCollector.zip)

**Comment by jrouwe** [comment]
B.t.w. @mihe what program are you using for the nice diagrams you always draw?

**Comment by mihe** [comment]
> B.t.w. @mihe what program are you using for the nice diagrams you always draw?

That would be [Excalidraw](https://excalidraw.com/). :)

**Comment by jrouwe** [comment]
Cool, thanks for the tips!

**Comment by Gamepro5** [comment]
> I haven't done much testing with move_and_slide(), so either it shares the same issue but we haven't found it, or move_and_slide() does something to curate the normals received from the move_and_collides().

[moveandslidebug.zip](https://github.com/jrouwe/JoltPhysics/files/12516800/moveandslidebug.zip)

I seem to have found a situation where `move_and_slide()` also suffers from this. I guess the answer was most of the time, move_and_slide() is less sensitive to this stuff but it can still affect it, so in other words a combination of both theories.

> You'll see that it uses the surface normal (the thing you want) to determine if a slope is too steep, but it uses the contact normal (mContactNormal, which is the thing you get through godot-jolt right now) for other things.

Maybe I can make a gdnative feature request to allow other things not previously exposed to be exposed. Maybe the surface normal would fix all these issues.

**Comment by mihe** [comment]
> I seem to have found a situation where move_and_slide() also suffers from this.

I don't quite understand what this repro is trying to show. The normals you print look fine, and the character doesn't seem to have any problems moving when setting `position` in `_ready()` instead, nor when switching to the default template script and just moving around. On top of that you're fetching the normals from `move_and_collide`, as opposed to `get_floor_normal`. There's also `get_slide_collision`, which lets you "open the hood" and see the actual `move_and_collide` data that was used in `move_and_slide`. That one will of course show you many of the same normals you've been seeing, but unless traversal is affected I don't see the problem.

Frankly, I would rather not see that this issue becomes re-framed to also include issues with `move_and_slide`. We've already concluded that not much can be done from a Jolt perspective here, so any issues where `move_and_slide` performs worse (in terms of traversal) when using Godot Jolt compared to Godot Physics should be filed in the Godot Jolt repo instead.

> Maybe I can make a gdnative feature request to allow other things not previously exposed to be exposed.

The contact data (much like the rest of the physics interface) is shared between Godot Physics and any physics server exposed from GDExtension (which is GDNative's successor in Godot 4), so this wouldn't be about changing the extension interface but rather changing the physics interface, which involves adding support for this in Godot Physics as well. Requesting that surface normals be added to Godot Physics would mean that it becomes a thing in Godot Jolt as well eventually.

It would technically be possible to add custom versions of `get_rest_info` and `body_test_motion` to Godot Jolt that bundle the surface normals as part of the contact data, but besides being a somewhat messy and unattractive solution, it also brings along other limitations, like not being accessible from C#, so I would much rather see that such a thing propagates through the shared physics interface instead.

**Comment by Gamepro5** [comment]
Even if position is set in _ready(), the move_and_slide() stops farther up the slope. The purpose of that was to just show that move_and_slide() also does suffer from this to some extent, although it's nowhere near as bad as mine.

> Frankly, I would rather not see that this issue becomes re-framed to also include issues with move_and_slide

Yeah, move_and_slide() is way more complex than move_and_collide(). I didn't expect a debug from that, I was just pointing out that our theory that `move_and_slide()` is unaffected by this isn't 100% accurate.

**Comment by mihe** [comment]
> Even if position is set in _ready(), the move_and_slide() stops farther up the slope.

I'll admit I was in too much of a hurry to let that one play out in full, so I did miss that one.

> I was just pointing out that our theory that move_and_slide() is unaffected by this isn't 100% accurate.

The problem is we've intertwined 3-4 separate issues into this whole thing and grouped them together as "bad/weird normals". `move_and_slide` has always been susceptible to triangle edge collisions to some degree, and is something that happens with Godot Physics as well. The fix for the [original issue](https://github.com/jrouwe/JoltPhysics/issues/630#issue-1821197791) here, which prompted the change of this "active edge threshold" from 5 to 50 degrees, fixed [one such problem](https://github.com/godot-jolt/godot-jolt/issues/524) with `move_and_slide`. I've resolved/lessened [another one](https://github.com/godot-jolt/godot-jolt/issues/529) during the time we've been looking into this one, and there's another similar fix in the pipeline.

What `move_and_slide` doesn't seem to have any problems with however are the normals generated by Jolt's active edge detection (aka `EActiveEdgeMode::CollideOnlyWithActive`), meaning the issue that prompted the reboot of this issue in the first place. The inability to "fix" these sort of poisoned the whole batch here and prompted the closing of this issue.

The issue your latest repro ran into looks to be another instance of scenario 2 mentioned [above](https://github.com/jrouwe/JoltPhysics/issues/630#issuecomment-1694669208), where the use of `mActiveEdgeMovementDirection` ends up having a less-than-ideal effect. I hadn't seen this manifest with `move_and_slide` yet, but I also do most of my testing with the more traditional `CapsuleShape3D`, which I would imagine is more resilient to this. I might experiment with the compromise I mentioned [above](https://github.com/jrouwe/JoltPhysics/issues/630#issuecomment-1704083749) and just omit `mActiveEdgeMovementDirection`, which fixes this particular problem. I get the sense that `mActiveEdgeMovementDirection` might not be a great fit for `move_and_collide`, but I'll have to do a lot more testing to make sure.

**Comment by Gamepro5** [comment]
is `mActiveEdgeMovementDirection` the one that makes the following situation report both normals? I remember when you made the minimum active angle bigger this suddenly became an issue, but that it wasn't important because it would get fixed after this issue gets fixed (because it wont get fixed).

Orange arrow is movement vector and other arrows are both normals that get reported form this instead of just up. It's not a great drawing.
![Untitled](https://github.com/jrouwe/JoltPhysics/assets/23440295/554af403-3eb5-4e88-a09e-4144703838d1)

> The reason you're seeing a difference between 0.5.0 and 0.6.0 is because the threshold angle for this behavior being increased from 5 degrees to 50 degrees, in order to fix https://github.com/godot-jolt/godot-jolt/issues/508 (and to some degree https://github.com/godot-jolt/godot-jolt/issues/524). As mentioned already this will negatively impact your particular character controller, but benefits RigidBody3D greatly, as well as move_and_slide. The proposed algorithm in https://github.com/jrouwe/JoltPhysics/issues/630 should resolve this particular problem for you.

This is a quote from https://github.com/godot-jolt/godot-jolt/issues/536#issuecomment-1697353808 . This doesn't seem to be an issue with `move_and_collide()` (i assume) because it doesn't care about what normals it gets from the snap check, it likely only uses movements from the actual movement trace that comes after.
